subsys/bluetooth/host/settings.c

```diff
@@ -298,6 +301,133 @@ int bt_settings_init(void)
 	return 0;
 }
 
+static const char * const bt_settings_keys[] = {
+	"sc",
+	"cf",
+	"ccc",
+	"hash",
+	"name",
+	"id",
+	"irk",
+	"link_key",
+	"keys",
+	"mesh",
+};
+
+static bool mem_eq(const uint8_t *a1, size_t n1, const uint8_t *a2, size_t n2)
+{
+	if (n1 != n2) {
+		return false;
+	}
+
+	return (memcmp(a1, a2, n1) == 0);
+}
+
+static int bt_settings_load_direct_cleanup(const char *key, size_t len, settings_read_cb read_cb,
+					   void *cb_arg, void *param)
+{
+	int err;
+	int next_len;
+	int subsys_len;
+	unsigned long id;
+	const char *next;
+	const char *id_str;
+	bool key_is_supported;
+	char full_key[BT_SETTINGS_KEY_MAX];
+
+	bt_addr_le_t *id_addr = (bt_addr_le_t *)param;
+
+	subsys_len = settings_name_next(key, &next);
+
+	LOG_DBG("subsys: %.*s", subsys_len, key);
+
+	key_is_supported = false;
+
+	for (size_t i = 0; i < ARRAY_SIZE(bt_settings_keys); i++) {
+		if (mem_eq(key, subsys_len, bt_settings_keys[i], strlen(bt_settings_keys[i]))) {
+			key_is_supported = true;
+		}
+	}
+
+	if (key_is_supported) {
+		if (!next || !strncmp(key, "mesh", subsys_len)) {
+			/* key exist but does not need to be checked for
+			 * leftover because it is unique and not linked to ID.
+			 * We return and continue to look for next keys.
+			 */
+			return 0;
+		}
+
+		next_len = settings_name_next(next, &id_str);
+
+		if (!id_str) {
+			id = BT_ID_DEFAULT;
+		} else {
+			id = strtoul(id_str, NULL, 10);
+
+			if (id >= CONFIG_BT_ID_MAX) {
+				LOG_ERR("Invalid local identity (%lu)", id);
+
+				return 0;
```

Comment:
Seems odd to return `0` and log something as an error. Intentional?

Reply:
Yes, this is intentional. If we return a non-zero value, settings subtree searching will be stopped, but I think we want to continue to clean the settings even if this error occur.

Reply:
Might be worth a in-code comment :)